torch/nested/_internal/cached_tensor.py

```diff
@@ -0,0 +1,131 @@
+from contextlib import contextmanager
+from typing import *  # noqa: F403
+
+import torch
+from torch.utils import _pytree as pytree
+
+
+class CachedTensor(torch.Tensor):
+    metadata: Dict[str, Optional[torch.Tensor]]
+    source_field: str
+
+    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
+    # is determined by a "source" tensor in the dict (specified by the user
+    # during construction).
+    #
+    # This class is not super useful on its own because by default, performing any
+    # operations on it will be as if you first unwrapped the CachedTensor and then
+    # performed the operation on the plain source tensor (a plain tensor is also returned).
+    # To leverage the extra metadata, you must register an op to perform the special logic
+    # you want via register_cached_tensor_func.
+    #
+    # When used this way (1) it is a convenient way to keep around metadata
+    # related to a tensor, without having to laboriously thread those extra metadata
+    # around, e.g. through aten signatures. (2) allows one to trigger custom
+    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
+    # otherwise because the subclass constructor op itself usually does not take the
+    # subclass itself as input!). See NestedTensor for an example.
+
+    @staticmethod
+    @torch._disable_dynamo  # type: ignore[misc]
```

Comment:
why does this need to be here?

Reply:
From talking with @bdhirsh its standard practice to prevent dynamo from randomly tracing this frame and erroring out in a weird way. (It's also used in DTensor.)

Reply:
This reminds me to add a test for constructing CachedTensor in the middle of the graph (which we may rely on for NestedTensor)